source/simulator/solver_schemes.cc

```diff
@@ -1069,6 +1070,124 @@ namespace aspect
     if (parameters.run_postprocessors_on_nonlinear_iterations)
       postprocess ();
   }
+
+
+
+  template <int dim>
+  void Simulator<dim>::solve_stokes_adjoint ()
+  {
+    // TODO iterate
+    // todo make number of interations a parameter and change Do iterations for inversion parameter
+//    for (unsigned int i=0; i<parameters.num_it_adjoint; ++i)
+//      {
+
+//        pcout << " ^^ Adjoint iteration number " << i  << std::endl;
+
+    // -------------------------------------------------------------
+    // SOLVE FORWARD PROBLEM
+
+    pcout << "   Forward problem ... " << std::endl;
+
+    // I think at this point this flag only affects postprocessing
+    adjoint_problem = false;
+
+    // set the assemblers for the RHS
+    set_assemblers();
+
+    rebuild_stokes_matrix = rebuild_stokes_preconditioner = true;
+
+    assemble_stokes_system();
+    build_stokes_preconditioner();
+
+    solve_stokes(); // solves Ax=b, puts x into 'solution'
+    // put 'solution' into 'current_linearization_point' (=u/p)
+    current_linearization_point.block(introspection.block_indices.velocities)
+      = solution.block(introspection.block_indices.velocities);
+    if (introspection.block_indices.velocities != introspection.block_indices.pressure)
+      current_linearization_point.block(introspection.block_indices.pressure)
+        = solution.block(introspection.block_indices.pressure);
+
+    // postprocess forward solution to calculate DT and maybe geoid etc.
+    // TODO: only run the postprocessors I need, e.g. DT
+    postprocess ();
+
+    // -------------------------------------------------------------
+    // SOLVE ADJOINT PROBLEM
+
+    pcout << "   Adjoint problem ... " << std::endl;
+
+    // only do adjoint if final (refined) mesh is reached
+    // I think at this point this flag only affects postprocessing
+    adjoint_problem = true;
+
+    // clear all RHS assemblers and only do pressure rhs compatibility and assemble the Adjoint RHS;
+    assemblers->stokes_system.clear();
+    assemblers->stokes_system_on_boundary_face.clear();
+
```

Comment:
You mentioned you need the pressure rhs modification? If so you can add it like this:
```
    // add the terms necessary to normalize the pressure
    if (do_pressure_rhs_compatibility_modification)
      assemblers->stokes_system.push_back(
        std_cxx14::make_unique<aspect::Assemblers::StokesPressureRHSCompatibilityModification<dim> >());

```
probably also needs SimulatorAccess like your assembler below